monitor/event_stats.hpp

```diff
@@ -0,0 +1,244 @@
+
+namespace ioremap { namespace monitor {
+
+class mutex_lock_policy
+{
+	typedef std::mutex mutex_t;
+public:
+	class unique_lock {
+	public:
+		unique_lock(mutex_lock_policy *policy)
+		: lock_(policy->get_lock())
+		{}
+	private:
+		std::unique_lock<mutex_t> lock_;
+	};
+
+	mutex_t &get_lock() { return mut_; }
+private:
+	mutex_t mut_;
+};
+
+class null_lock_policy
+{
+public:
+	class unique_lock {
+	public:
+		unique_lock(null_lock_policy *policy) { (void) policy; }
+	};
+};
+
+/*
+ *	This class provides statistics of events during some defined period of time.
+ * Also there are strict limitation of memory resources which could be used by
+ * instances of this class. So some information about events may be lost, if
+ * significant amount of events multiplied by event size exceeded memory limitations.
+ * Therefore statistics are computed approximately in case of lots of heavyweight events.
+ *	Any type E of event must support following methods: (get/set)_weight, (get/set)_frequency,
+ * and (get/set)_time.
+ *	Currently only get_top method implemented, which could be used for getting top k events
+ * with highest weight, where weight is approximate sum of weight of particular events during
+ * observable period of time (period_in_seconds).
+ */
+template<typename E, typename lock_policy = null_lock_policy>
+class event_stats : private lock_policy
+{
+	// events stored in fixed-size LRU that is implemented using treap
+	typedef ioremap::cache::treap<E> treap_t;
+public:
+	/*!
+	 * \internal
+	 *
+	 * Constructor parameters: \a events_size - maximum memory available for internal data structures,
+	 * \a period_in_seconds - observable period of time
+	 */
+	event_stats(size_t events_size, int period_in_seconds)
+	: num_events(0),
+	 max_events(bytes_to_num_events(events_size)),
+	 period(period_in_seconds)
+	{}
+
+	~event_stats() {
+		clear();
+	}
+
+	/*!
+	 * \internal
+	 *
+	 * Add \a event with \a time occurence of this event to update statistics
+	 */
+	void add_event(const E &event, time_t time)
+	{
+		typename lock_policy::unique_lock lock(this);
+		auto it = treap.find(event.get_key());
+		if (it) {
+			update_weight(*it, time, period, event.get_weight());
+			update_frequency(*it, time, period, 1.);
+			it->set_time(time);
+			treap.decrease_key(it);
+		} else {
+			if (num_events < max_events) {
+				treap.insert(new E(event));
+				++num_events;
+			} else {
+				auto t = treap.top();
+				treap.erase(t);
+				*t = event;
+				treap.insert(t);
+			}
+		}
+	}
+
+	/*!
+	 * \internal
+	 *
+	 * Get top \a k events with highest weight, \a time - current time,
+	 * \a top_size - result container, that should support push_back operation and random access iterators
+	 */
+	template< typename ResultContainer >
+	void get_top(size_t k, time_t time, ResultContainer &top_size)
+	{
+		std::vector< typename treap_t::p_node_type > top_nodes;
+
+		{
+			typename lock_policy::unique_lock lock(this);
+			treap_to_container(treap.top(), top_nodes);
```

Comment:
First of all I must mention, that event_stats designed as data structure with frequent add_event and rare get_top operations access pattern. So add_event must be as fast as possible to not affect overall performance.

Complexity of this operation (get_top) depends on number of expired events (events with timestamp value outside window size) - E, and total number of events - N.
E <= N

It does some operations under lock:
> O(N) - treap_to_container()
> O(E * logN) - worst case for sequental treap.erase()

And after unlock:
> std::partial_sort() // standart guarantees O(N * log k)
It is also possible to make heap of k elements and do (N-k) heap.update_min() operations, resulting in O(k + (N-k)*log k) = O((N-k)*log k), but with a small k complexity of this method is not significantly faster than std::partial_sort().